File: frontend/lib/education.ts
export type Lesson = {
  slug: string;
  title: string;
  description: string;
  level: string;
  competency: string;
  estimated_minutes: number;
  prerequisites: string[];
  status: string;
  score: number | null;
  completed_at: string | null;
  path_state: "available" | "recommended" | "locked" | "completed" | "remediation";
  path_reason: string;
};

export type CompetencySummary = {
  competency: string;
  completed_lessons: number;
  total_lessons: number;
  assessed_lessons: number;
  average_score: number | null;
  mastery_percent: number;
  proficiency_band: string;
};

export type CompetencyTrend = {
  competency: string;
  attempt_count: number;
  recent_average: number | null;
  prior_average: number | null;
  direction: "improving" | "stable" | "declining" | "insufficient_evidence";
  explanation: string;
};

export type EducationHubData = {
  completed_lessons: number;
  total_lessons: number;
  assessed_lessons: number;
  completion_percent: number;
  average_score: number | null;
  mastery_percent: number;
  proficiency_band: string;
  recommended_lesson_slug: string | null;
  recommendation_reason: string;
  lessons: Lesson[];
  competencies: CompetencySummary[];
};

export type AssessmentQuestion = {
  id: string;
  prompt: string;
  options: string[];
  objective: string;
};

export type AdaptiveAssessment = {
  lesson_slug: string;
  competency: string;
  difficulty: string;
  difficulty_reason: string;
  question: AssessmentQuestion;
};

export type AssessmentAttempt = {
  id: number;
  lesson_slug: string;
  competency: string;
  difficulty: string;
  question_id: string;
  selected_option: number;
  score: number;
  passed: boolean;
  created_at: string;
};

export type AssessmentResult = {
  lesson_slug: string;
  competency: string;
  difficulty: string;
  score: number;
  passed: boolean;
  feedback: string;
  learning_objective: string;
  education_hub: EducationHubData;
};

File: backend/app/schemas/education.py
from datetime import datetime

from pydantic import BaseModel, Field, model_validator


MASTERY_SCORE_THRESHOLD = 70


class LessonProgressUpdate(BaseModel):
    status: str = Field(pattern="^(not_started|in_progress|completed)$")
    score: int | None = Field(default=None, ge=0, le=100)

    @model_validator(mode="after")
    def require_passing_score_for_completion(self) -> "LessonProgressUpdate":
        if self.status == "completed" and (self.score is None or self.score < MASTERY_SCORE_THRESHOLD):
            raise ValueError(
                f"Completed lessons require a score of at least {MASTERY_SCORE_THRESHOLD}%."
            )
        return self


class LessonRead(BaseModel):
    slug: str
    title: str
    description: str
    level: str
    competency: str
    estimated_minutes: int
    prerequisites: list[str]
    status: str
    score: int | None
    completed_at: datetime | None
    path_state: str
    path_reason: str


class CompetencySummary(BaseModel):
    competency: str
    completed_lessons: int
    total_lessons: int
    assessed_lessons: int
    average_score: int | None
    mastery_percent: int
    proficiency_band: str


class EducationHubRead(BaseModel):
    completed_lessons: int
    total_lessons: int
    assessed_lessons: int
    completion_percent: int
    average_score: int | None
    mastery_percent: int
    proficiency_band: str
    recommended_lesson_slug: str | None
    recommendation_reason: str
    lessons: list[LessonRead]
    competencies: list[CompetencySummary]


class AssessmentQuestionRead(BaseModel):
    id: str
    prompt: str
    options: list[str]
    objective: str


class AdaptiveAssessmentRead(BaseModel):
    lesson_slug: str
    competency: str
    difficulty: str
    difficulty_reason: str
    question: AssessmentQuestionRead


class AssessmentSubmission(BaseModel):
    question_id: str
    selected_option: int = Field(ge=0)


class AssessmentResultRead(BaseModel):
    lesson_slug: str
    competency: str
    difficulty: str
    score: int
    passed: bool
    feedback: str
    learning_objective: str
    education_hub: EducationHubRead
